### ISSUE TITLE
pnpm doesn't recognise cached/installed version of node and keeps downloading it

### ISSUE THREAD

**Issue by antony** [pnpm/pnpm#6926]
This is costing us hours of wasted time per week, so I'm raising it in the company discussion thread too.

### pnpm version:

8.6.2

### Code to reproduce the issue:

seems to happen on github actions exclusively. simply running:

```
pnpm i
```

We cache the node version downloaded by running the following:

```
      - uses: actions/cache@v3
        name: cache node
        with:
          path: /home/runner/.local/share/pnpm
          key: ${{ runner.os }}-pnpm-node-cache
          restore-keys: |
            ${{ runner.os }}-pnpm-node-cache-
```

We used to use a pnpm command (I forget which) to determine the cache dir name, but this in itself caused pnpm to download node (!!!) and thus it would stick even before the caching happened. So now we hardcode it, based on looking at where pnpm installs node to.

### Expected behavior:

PNPM sees that node is installed on the machine, and doesn't re-download it.

### Actual behavior:

PNPM insists on downloading node from nodejs.org which frequently returns gateway errors or times out

![Screenshot from 2023-08-09 12-20-36](https://github.com/pnpm/pnpm/assets/218949/3182d853-9a31-44d0-90c2-584b1b29cd5a)

### Additional information:

 - `node -v` prints:
v19.9.0
 - Windows, macOS, or Linux?:
 - Ubuntu Linux

**Comment by KSXGitHub** [comment]
What is the result of `pnpm config list | grep node-version`?

**Comment by KSXGitHub** [comment]
BTW, in the github action, your `restore-keys` doesn't match the `key`. You should remove it.

**Comment by antony** [comment]
❯ npm config list | grep -i node-version
node-version = "18.17.0" 
use-node-version = "18.17.0" 
❯ grep -i node-version < .npmrc
use-node-version=18.17.0
node-version=18.17.0

I've removed the restore key - thanks :)

**Comment by KSXGitHub** [comment]
@antony You should remove `use-node-version` from `.npmrc`, either from your repo, or by patching them during CI.

If you insist on keeping `use-node-version`, I suggest try running `pnpm exec which node` to find the path to Node.js and add its directory to cache. Although, keeping `use-node-version` would render your CI matrix of Node.js versions useless.

@zkochan Is there a way to ignore `use-node-version` during CI only?

**Comment by antony** [comment]
@KSXGitHub I'm a bit confused. If I remove use-node-version, then pnpm is no longer managing the node version used on CI. This means that CI would use a different node version than the one we configure for the project. Thus, I feel like your fix makes my problem go away, but only by disabling the feature?

We don't have a matrix of node versions on CI, it's not an open source project - it's our own software, which we only need to run and be tested against a single node version - the same one we use on our hosting environment, locally, and on CI.

**Comment by KSXGitHub** [comment]
> If I remove use-node-version, then pnpm is no longer managing the node version used on CI. This means that CI would use a different node version than the one we configure for the project. Thus, I feel like your fix makes my problem go away, but only by disabling the feature?

Correct. `pnpm` would no longer download Node.js and would use the system's Node.js.

> We don't have a matrix of node versions on CI, it's not an open source project - it's our own software, which we only need to run and be tested against a single node version - the same one we use on our hosting environment, locally, and on CI.

So you insist on letting pnpm managing your Node.js versions? Then I suggest:
1. Add `~/.local/share/pnpm/nodejs/18.17.0/bin/` to the `path` of your cache action.
2. (Optional) You should set [`standalone`](https://github.com/pnpm/action-setup/tree/v2.4.0#standalone) in your `pnpm/action-setup` to `true`. This would make pnpm be independent from local Node.js.
3. (Optional) You may replace `/home/runner` in your cache action with `~`.

**Comment by KSXGitHub** [comment]
> We don't have a matrix of node versions on CI, it's not an open source project - it's our own software, which we only need to run and be tested against a single node version - the same one we use on our hosting environment, locally, and on CI.

`PNPM_HOME` is redefined to be in `node_modules/.bin`. Which is why your cache didn't work.

**Comment by antony** [comment]
> PNPM_HOME is redefined to be $dest/node_modules/.bin (dest is ~/setup-pnpm by default). Which is why your cache didn't work.

ok - $PNPM_HOME is what I was looking for.

If I run it locally though, it points to where I expect it to point, not where you're suggesting it would be:

❯ echo $PNPM_HOME
/home/ant/.local/share/pnpm

on CI, running `ls` on that same directory (/home/runner/.local/share/pnpm) shows all of the downloaded node installs, which is how I found it. The cache indeed caches about 33mb of data, which is around the size of a node install.

Locally, 

```js
ls ~/setup-pnpm
ls: cannot access '/home/ant/setup-pnpm': No such file or directory
```

I assume you mean `~/.setup-pnpm` which also doesn't exist:

```js
ls ~/.setup-pnpm
ls: cannot access '/home/ant/.setup-pnpm': No such file or directory
```

I will echo out $PNPM_HOME on CI and see what it resolves to.

**Comment by KSXGitHub** [comment]
> If I run it locally though



> Locally,
> 
> ```js
> ls ~/setup-pnpm
> ls: cannot access '/home/ant/setup-pnpm': No such file or directory
> ```

Sorry for not being clear, I meant in GitHub CI. And it's `~/setup-pnpm` (no dot).

The relevant code: https://github.com/pnpm/action-setup/blob/d882d12c64e032187b2edb46d3a0d003b7a43598/src/install-pnpm/run.ts#L31

**Comment by antony** [comment]
It looks like the $PNPM_HOME variable is set by action-setup, which, apologies, I wasn't clear - I don't use.

I instead use corepack, so corepack handles pnpm for me, and sets no such variable:

![Screenshot from 2023-08-10 11-47-33](https://github.com/pnpm/pnpm/assets/218949/a6554b65-2359-450e-8443-1bada1c97823)

However, are you suggesting that if I set that variable myself, I can tell PNPM to use wherever I like as it's home? If so, that may indeed solve my issue.

**Comment by KSXGitHub** [comment]
> However, are you suggesting that if I set that variable myself, I can tell PNPM to use wherever I like as it's home? If so, that may indeed solve my issue.

I think yeah, it should work.

> I instead use corepack, so corepack handles pnpm for me

So you use Node.js's corepack to manage pnpm AND pnpm's `.npmrc` to manage Node.js at the same time? I don't think I would want to do that.

**Comment by antony** [comment]
Corepack can only choose my package manager. I want everyone on my team to use the same exact version of pnpm without having to think about it.
Then, I want everybody on my team to use the same node version, so I let pnpm handle that, also without any effort or setup from my team outside `corepack enable`.

We used to use volta but it has some pain points - this works well other than the bug I raised around putting random props in `.npmrc`

what makes you say that this is a bad idea?

**Comment by KSXGitHub** [comment]
Nothing in particular. I worried too much.

BTW, have you successfully cache Node.js?

**Comment by antony** [comment]
Still working on it - unfortunately a very busy day as usual! I'll let you know asap.

**Comment by antony** [comment]
It doesn't work. Despite setting PNPM_HOME env var to ~/setup-pnpm, pnpm still tries to download node, succeeds in downloading it, but the post-cache job tells me that ~/setup-pnpm directory doesn't exist, so nothing is cached.

**Comment by KSXGitHub** [comment]
So I created a [test workflow](https://github.com/KSXGitHub/pnpm-cache-node-env-6926/actions/runs/5874832522/job/15930221793) and the cache works as expected:

![cache screenshot](https://github.com/pnpm/pnpm/assets/11488886/dbf7a02c-294c-462d-a883-aed0ce318b90)

[(repo)](https://github.com/KSXGitHub/pnpm-cache-node-env-6926)

There must be something else going on with your workflow.

**Comment by KSXGitHub** [comment]
In the workflow I created, the first attempt fetched Node.js as expected:

![screenshot of attempt #1](https://github.com/pnpm/pnpm/assets/11488886/0c56c6b9-e469-4bcb-86a5-c2b204eec257)

In the second attempt (created by pressing "Re-run all jobs" button), there is no such `Fetching Node.js 18.17.0` message because there is already a cache:

![screenshot of attempt #2](https://github.com/pnpm/pnpm/assets/11488886/f9a855ae-eeb7-4d16-b98f-2e0aada816a6)

**Comment by antony** [comment]
HI @KSXGitHub - I think you're right.

I changed my caching from using an absolute path to a homedir prefixed path and I no longer see the `Fetching...` line. Oddly, I did do this in the course of trying to raise this issue, and the issue was not resolved, so I'm not entirely sure why this now works.

Either way, it appears somehow that I'm now managing to cache the node install, and pnpm is behaving as I expect.

Thank you so much for your effort in helping me diagnose this.
